File: results/hc7_exact7_two_lobe_state_probe.cpp
#include <array>
#include <cstdint>
#include <iostream>
#include <vector>

// Exhaustive seven-label probe for the two-lobe state mechanism.
// Vertices are 0,...,6.  A,B are the literal label sets contacted by the
// two lobes.  We test whether two adjacent carriers with contact sets A,B
// fund an admissible two-block state as in the audited carrier lemma.

static int ei[7][7];

static bool is_clique(int q, uint32_t e) {
  for (int i=0;i<7;i++) if (q&(1<<i))
    for (int j=i+1;j<7;j++) if (q&(1<<j))
      if (!(e&(1u<<ei[i][j]))) return false;
  return true;
}

static bool is_independent(int a, uint32_t e) {
  for (int i=0;i<7;i++) if (a&(1<<i))
    for (int j=i+1;j<7;j++) if (a&(1<<j))
      if (e&(1u<<ei[i][j])) return false;
  return true;
}

static bool sees_block(int q, int block, uint32_t e) {
  if (block&(1<<q)) return true;
  for (int x=0;x<7;x++) if ((block&(1<<x)) && (e&(1u<<ei[q][x]))) return true;
  return false;
}

static bool two_state(int A, int B, uint32_t e) {
  // Q is a literal clique; I,J are nontrivial independent blocks.
  for (int Q=0;Q<128;Q++) {
    if (!is_clique(Q,e)) continue;
    int rem=127^Q;
    for (int I=rem;I;I=(I-1)&rem) {
      int J=rem^I;
      if (__builtin_popcount((unsigned)I)<2 || __builtin_popcount((unsigned)J)<2) continue;
      if ((I&~A) || (J&~B)) continue;
      if (!is_independent(I,e) || !is_independent(J,e)) continue;
      bool ok=true;
      for (int q=0;q<7;q++) if (Q&(1<<q)) {
        if (!((A&(1<<q)) || sees_block(q,I,e))) ok=false;
        if (!((B&(1<<q)) || sees_block(q,J,e))) ok=false;
      }
      if (ok) return true;
    }
  }
  return false;
}

static bool one_state(uint32_t e) {
  for (int I=1;I<128;I++) {
    if (__builtin_popcount((unsigned)I)<2 || !is_independent(I,e)) continue;
    int Q=127^I;
    if (is_clique(Q,e)) return true;
  }
  return false;
}

int main() {
  int k=0;
  for (int i=0;i<7;i++) for (int j=i+1;j<7;j++) ei[i][j]=ei[j][i]=k++;
  std::vector<uint32_t> tris;
  for (int i=0;i<7;i++) for (int j=i+1;j<7;j++) for (int l=j+1;l<7;l++)
    tris.push_back((1u<<ei[i][j])|(1u<<ei[i][l])|(1u<<ei[j][l]));

  long long checked=0, bad=0;
  for (uint32_t e=0;e<(1u<<21);e++) {
    bool tf=true; for (auto t:tris) if ((e&t)==t) {tf=false;break;}
    if (!tf || one_state(e)) continue;
    for (int A=0;A<128;A++) if (__builtin_popcount((unsigned)A)>=4)
      for (int B=0;B<128;B++) if (__builtin_popcount((unsigned)B)>=4 && (A|B)==127) {
        checked++;
        if (!two_state(A,B,e) && !two_state(B,A,e)) {
          bad++;
          if (bad<=40) {
            std::cout << "counterexample e="<<e<<" A="<<A<<" B="<<B
                      <<" sizes="<<__builtin_popcount((unsigned)A)<<","<<__builtin_popcount((unsigned)B)
                      <<" overlap="<<__builtin_popcount((unsigned)(A&B))<<"\n";
            std::cout << "edges";
            for(int i=0;i<7;i++)for(int j=i+1;j<7;j++)if(e&(1u<<ei[i][j]))std::cout<<" "<<i<<j;
            std::cout<<"\n";
          }
        }
      }
  }
  std::cout << "checked="<<checked<<" bad="<<bad<<"\n";
}

File: results/hc7_exact7_two_lobe_state_probe.py
#!/usr/bin/env python3
"""Falsify/test a boundary-only two-lobe state principle.

Enumerates labelled triangle-free graphs H on S=[7] with no admissible
one-block partition, and contact sets A,B of size at least four.  It asks
whether the anticomplete lobe carriers with contacts A,B can fund an
admissible two-block state via Lemma 3.1 of the thin-shore exchange.

This deliberately ignores gate portals.  A counterexample therefore
shows that lobe contact existence alone cannot close the c=2 gate cell.
"""

from itertools import combinations

N = 7
ALL = (1 << N) - 1
PAIRS = [(i, j) for i in range(N) for j in range(i + 1, N)]


def is_clique(mask, adj):
    vs = [i for i in range(N) if mask >> i & 1]
    return all(adj[u] >> v & 1 for u, v in combinations(vs, 2))


def is_independent(mask, adj):
    return all((adj[v] & mask) == 0 for v in range(N) if mask >> v & 1)


def no_one_block(adj):
    for q in range(1 << N):
        i = ALL ^ q
        if i.bit_count() >= 2 and is_independent(i, adj) and is_clique(q, adj):
            return False
    return True


def has_two_block_state(adj, A, B, carriers_adjacent=False):
    # Ordered I1 for A, I2 for B; Q is the remainder.
    sub = A
    while sub:
        I1 = sub
        sub = (sub - 1) & A
        if I1.bit_count() < 2 or not is_independent(I1, adj):
            continue
        available = B & ~I1
        sub2 = available
        while sub2:
            I2 = sub2
            sub2 = (sub2 - 1) & available
            if I2.bit_count() < 2 or not is_independent(I2, adj):
                continue
            Q = ALL ^ (I1 | I2)
            if not is_clique(Q, adj):
                continue
            # If the carriers are anticomplete, a boundary edge must supply
            # adjacency after the independent blocks are absorbed.
            if (not carriers_adjacent and
                    not any(adj[v] & I2 for v in range(N) if I1 >> v & 1)):
                continue
            ok = True
            for q in range(N):
                if not (Q >> q & 1):
                    continue
                if not (A >> q & 1) and not (adj[q] & I1):
                    ok = False
                if not (B >> q & 1) and not (adj[q] & I2):
                    ok = False
            if ok:
                return (I1, I2, Q)
    return None


def main():
    contact_sets = [m for m in range(1 << N) if m.bit_count() >= 4]
    triangle_masks = []
    for tri in combinations(range(N), 3):
        em = 0
        for uv in combinations(tri, 2):
            em |= 1 << PAIRS.index(tuple(sorted(uv)))
        triangle_masks.append(em)

    checked_h = checked_pairs = 0
    for emask in range(1 << len(PAIRS)):
        if any((emask & tm) == tm for tm in triangle_masks):
            continue
        adj = [0] * N
        for k, (u, v) in enumerate(PAIRS):
            if emask >> k & 1:
                adj[u] |= 1 << v
                adj[v] |= 1 << u
        if not no_one_block(adj):
            continue
        checked_h += 1
        for A in contact_sets:
            for B in contact_sets:
                checked_pairs += 1
                if (A | B) != ALL:
                    continue
                if has_two_block_state(adj, A, B, carriers_adjacent=True) is None:
                    print("COUNTEREXAMPLE")
                    print("edges", [PAIRS[k] for k in range(len(PAIRS)) if emask >> k & 1])
                    print("A", [v for v in range(N) if A >> v & 1])
                    print("B", [v for v in range(N) if B >> v & 1])
                    print("checked_h", checked_h, "checked_pairs", checked_pairs)
                    return
    print("NO COUNTEREXAMPLE", checked_h, checked_pairs)


if __name__ == "__main__":
    main()
